src/engine/execute.js

```diff
@@ -107,16 +107,161 @@ const handleReport = function (
+         * @type {boolean}
+         */
+        this._isHat = false;
+
+        /**
+         * The block opcode's implementation function.
+         * @type {?function}
+         */
+        this._blockFunction = null;
+
+        /**
+         * Is the block function defined for this opcode?
+         * @type {boolean}
+         */
+        this._definedBlockFunction = false;
+
+        /**
+         * Is this block a block with no function but a static value to return.
+         * @type {boolean}
+         */
+        this._isShadowBlock = false;
+
+        /**
+         * The static value of this block if it is a shadow block.
+         * @type {?any}
+         */
+        this._shadowValue = null;
+
+        /**
+         * A copy of the block's fields that may be modified.
+         * @type {object}
+         */
+        this._fields = Object.assign({}, this.fields);
+
+        /**
+         * A copy of the block's inputs that may be modified.
+         * @type {object}
+         */
+        this._inputs = Object.assign({}, this.inputs);
+
+        /**
+         * An arguments object for block implementations. All executions of this
+         * specific block will use this objecct.
+         * @type {object}
+         */
+        this._argValues = {
+            mutation: this.mutation
+        };
+
+        const {runtime} = blockUtility.sequencer;
+
+        const {opcode, fields, inputs} = this;
+
+        // Assign opcode isHat and blockFunction data to avoid dynamic lookups.
+        this._isHat = runtime.getIsHat(opcode);
+        this._blockFunction = runtime.getOpcodeFunction(opcode);
+        this._definedBlockFunction = typeof this._blockFunction !== 'undefined';
+
+        const fieldKeys = Object.keys(fields);
+
+        // Store the current shadow value if there is a shadow value.
+        this._isShadowBlock = fieldKeys.length === 1 && Object.keys(inputs).length === 0;
+        this._shadowValue = this._isShadowBlock && fields[fieldKeys[0]].value;
+
+        // Store the static fields onto _argValues.
+        for (const fieldName in fields) {
+            if (fieldName === 'VARIABLE') {
+                this._argValues.VARIABLE = {
+                    id: fields.VARIABLE.id,
+                    name: fields.VARIABLE.value
+                };
+            } else if (fieldName === 'LIST') {
+                this._argValues.LIST = {
+                    id: fields.LIST.id,
+                    name: fields.LIST.value
+                };
+            } else if (fieldName === 'BROADCAST_OPTION') {
+                this._argValues.BROADCAST_OPTION = {
+                    id: fields.BROADCAST_OPTION.id,
+                    name: fields.BROADCAST_OPTION.value
+                };
+            } else {
+                this._argValues[fieldName] = fields[fieldName].value;
+            }
+        }
+
+        // Remove custom_block. It is not part of block execution.
+        delete this._inputs.custom_block;
```

Comment:
I'm not sure I understand what's happening here.

Reply:
`custom_block` is information stored in inputs of a block that determines non-execution details. Before BlocksExecuteCache the for loop on inputs in execute tested skipped any `inputName === 'custom_block'` inputs. Every input of every executed block made that test. With BlocksExecuteCache we can make a clone of `inputs` and remove the ignored field for a performance gain.